File: pipeline-dashboard/server/discovery.js
'use strict';

const { DISCOVERY_RULES } = require('./config');
const { listFolderJobs } = require('./jenkins');

/** Compare two ganges version strings (e.g. "7.6.0.1") numerically, desc-friendly. */
function compareVersions(a, b) {
  const pa = a.split('.').map(Number);
  const pb = b.split('.').map(Number);
  const len = Math.max(pa.length, pb.length);
  for (let i = 0; i < len; i++) {
    const x = pa[i] || 0;
    const y = pb[i] || 0;
    if (x !== y) return x - y;
  }
  return 0;
}

/** Group version string into its "train" (major.minor), e.g. 7.6.0.1 -> 7.6 */
function trainOf(version) {
  const parts = version.split('.');
  return parts.slice(0, 2).join('.');
}

/**
 * Discover all version-scoped pipelines across the configured rules.
 * Returns a structure keyed by version, each holding the matching jobs, plus a
 * separate list of "master" jobs (one per rule/lane).
 *
 * Shape:
 * {
 *   masters: [ { key, controller, path, lane, label, title, subtitle } ],
 *   versions: {
 *      "7.6": [ { key, controller, path, lane, version, label, title, subtitle } ]
 *   },
 *   discoveredAt, errors: []
 * }
 */
async function discoverPipelines() {
  const masters = [];
  const versions = {};
  const errors = [];

  for (const rule of DISCOVERY_RULES) {
    const res = await listFolderJobs(rule.controller, rule.parent);
    if (!res.ok) {
      errors.push({ rule: rule.id, error: res.error });
      continue;
    }

    // Master job for this lane (some lanes, e.g. Precommit PC, have none).
    const masterJob = rule.masterName
      ? res.jobs.find((j) => j.name === rule.masterName)
      : null;
    if (masterJob) {
      masters.push({
        key: `${rule.id}:master`,
        controller: rule.controller,
        path: [...rule.parent, rule.masterName],
        lane: rule.lane,
        ruleId: rule.id,
        masterGroup: rule.masterGroup || 'other',
        label: rule.label,
        title: `${rule.shortLabel} master`,
        subtitle: masterJob.name,
        version: 'master',
      });
    }

    // Version-scoped jobs.
    for (const job of res.jobs) {
      const m = rule.versionRegex.exec(job.name);
      if (!m) continue;
      const version = m[1];
      if (!versions[version]) versions[version] = [];
      versions[version].push({
        key: `${rule.id}:${version}`,
        controller: rule.controller,
        path: [...rule.parent, job.name],
        lane: rule.lane,
        ruleId: rule.id,
        masterGroup: rule.masterGroup || 'other',
        label: rule.label,
        title: `${rule.shortLabel} ${version}`,
        subtitle: job.name,
        version,
      });
    }

    // Synthesize a master card from the newest version when the lane has no real
    // master job (e.g. Precommit PC). Represents the "mainline" precommit status.
    if (rule.synthesizeMasterFromLatest) {
      const matched = res.jobs
        .map((j) => {
          const m = rule.versionRegex.exec(j.name);
          return m ? { name: j.name, version: m[1] } : null;
        })
        .filter(Boolean)
        .sort((a, b) => compareVersions(b.version, a.version));
      const newest = matched[0];
      if (newest) {
        masters.push({
          key: `${rule.id}:master`,
          controller: rule.controller,
          path: [...rule.parent, newest.name],
          lane: rule.lane,
          ruleId: rule.id,
          masterGroup: rule.masterGroup || 'other',
          label: rule.label,
          title: `${rule.shortLabel} (latest ${newest.version})`,
          subtitle: newest.name,
          version: 'master',
          synthesizedFrom: newest.version,
        });
      }
    }
  }

  return {
    masters,
    versions,
    trainOf,
    compareVersions,
    discoveredAt: Date.now(),
    errors,
  };
}

/** Sorted version keys, newest first. */
function sortedVersions(versionsObj) {
  return Object.keys(versionsObj).sort((a, b) => compareVersions(b, a));
}

module.exports = { discoverPipelines, compareVersions, trainOf, sortedVersions };

File: pipeline-dashboard/go/internal/discovery/discovery.go
// Package discovery auto-finds version-scoped pipelines by listing Jenkins
// folders and regexing the version out of child job names.
// Ported from server/discovery.js.
package discovery

import (
	"sort"
	"strconv"
	"strings"
	"time"

	"github.com/nutanix/msp-pipeline-dashboard/internal/config"
	"github.com/nutanix/msp-pipeline-dashboard/internal/jenkins"
)

// Meta describes one pipeline to fetch (produced by config + discovery).
type Meta struct {
	Key             string
	Controller      string
	Path            []string
	Lane            string
	RuleID          string
	MasterGroup     string
	Label           string
	Title           string
	Subtitle        string
	Version         string
	Category        string
	SynthesizedFrom string
}

// Result is the discovery output.
type Result struct {
	Masters      []Meta
	Versions     map[string][]Meta
	DiscoveredAt int64
	ErrorList    []DiscErr
}

// DiscErr is a per-rule folder-listing failure.
type DiscErr struct {
	Rule  string
	Error string
}

// CompareVersions compares two ganges version strings numerically, segment by
// segment. Returns <0 if a<b, 0 if equal, >0 if a>b.
func CompareVersions(a, b string) int {
	pa := splitNums(a)
	pb := splitNums(b)
	n := len(pa)
	if len(pb) > n {
		n = len(pb)
	}
	for i := 0; i < n; i++ {
		var x, y int
		if i < len(pa) {
			x = pa[i]
		}
		if i < len(pb) {
			y = pb[i]
		}
		if x != y {
			return x - y
		}
	}
	return 0
}

func splitNums(v string) []int {
	parts := strings.Split(v, ".")
	out := make([]int, len(parts))
	for i, p := range parts {
		out[i], _ = strconv.Atoi(p)
	}
	return out
}

// TrainOf groups a version into its major.minor train, e.g. 7.6.0.1 -> 7.6.
func TrainOf(version string) string {
	parts := strings.Split(version, ".")
	if len(parts) > 2 {
		parts = parts[:2]
	}
	return strings.Join(parts, ".")
}

// DiscoverPipelines lists every discovery rule's folder and builds the meta set.
func DiscoverPipelines() Result {
	masters := []Meta{}
	versions := map[string][]Meta{}
	errs := []DiscErr{}

	for _, rule := range config.DiscoveryRules {
		res := jenkins.ListFolderJobs(rule.Controller, rule.Parent)
		if !res.OK {
			errs = append(errs, DiscErr{Rule: rule.ID, Error: res.Error})
			continue
		}

		// Master job for this lane, when it has one.
		if rule.MasterName != "" {
			for _, j := range res.Jobs {
				if j.Name == rule.MasterName {
					mg := rule.MasterGroup
					if mg == "" {
						mg = "other"
					}
					masters = append(masters, Meta{
						Key:         rule.ID + ":master",
						Controller:  rule.Controller,
						Path:        append(append([]string{}, rule.Parent...), rule.MasterName),
						Lane:        rule.Lane,
						RuleID:      rule.ID,
						MasterGroup: mg,
						Label:       rule.Label,
						Title:       rule.ShortLabel + " master",
						Subtitle:    j.Name,
						Version:     "master",
					})
					break
				}
			}
		}

		// Version-scoped jobs. Later rules win on the same version+lane so
		// controller-1 LKG replaces harbinger-14 for overlapping trains.
		for _, j := range res.Jobs {
			m := rule.VersionRegex.FindStringSubmatch(j.Name)
			if m == nil || len(m) < 2 {
				continue
			}
			version := m[1]
			mg := rule.MasterGroup
			if mg == "" {
				mg = "other"
			}
			upsertVersionJob(versions, Meta{
				Key:         rule.ID + ":" + version,
				Controller:  rule.Controller,
				Path:        append(append([]string{}, rule.Parent...), j.Name),
				Lane:        rule.Lane,
				RuleID:      rule.ID,
				MasterGroup: mg,
				Label:       rule.Label,
				Title:       rule.ShortLabel + " " + version,
				Subtitle:    j.Name,
				Version:     version,
			})
		}
	}

	return Result{
		Masters:      masters,
		Versions:     versions,
		DiscoveredAt: time.Now().UnixMilli(),
		ErrorList:    errs,
	}
}

// upsertVersionJob records a version-scoped job. A later rule with the same
// version+lane replaces the earlier one (used when LKG moved controllers).
func upsertVersionJob(versions map[string][]Meta, meta Meta) {
	list := versions[meta.Version]
	for i, existing := range list {
		if existing.Lane == meta.Lane {
			list[i] = meta
			versions[meta.Version] = list
			return
		}
	}
	versions[meta.Version] = append(list, meta)
}

// SortedVersions returns version keys newest-first.
func SortedVersions(versions map[string][]Meta) []string {
	keys := make([]string, 0, len(versions))
	for k := range versions {
		keys = append(keys, k)
	}
	sort.Slice(keys, func(i, j int) bool {
		return CompareVersions(keys[i], keys[j]) > 0
	})
	return keys
}
